pandas/core/dtypes/concat.py

```diff
@@ -89,14 +89,16 @@ def _get_series_result_type(result, objs=None):
 def _get_frame_result_type(result, objs):
     """
     return appropriate class of DataFrame-like concat
-    if any block is SparseBlock, return SparseDataFrame
+    if all blocks are SparseBlock, return SparseDataFrame
     otherwise, return 1st obj
     """
-    if any(b.is_sparse for b in result.blocks):
-        from pandas.core.sparse.api import SparseDataFrame
+
+    from pandas.core.sparse.api import SparseDataFrame
+
+    if result.blocks and all(b.is_sparse for b in result.blocks):
         return SparseDataFrame
     else:
-        return objs[0]
+        return next(obj for obj in objs if not type(obj) == SparseDataFrame)
```

Comment:
use ABCSparseDataFrame

Reply:
actually need to add this in pandas/core/dtypes/generic.py (and a test for it)

use  ``isinstance(obj, ABCSparseDataFrame)``

Reply:
hmm you can just return a ``DataFrame`` (this migth be subclasses though)

Reply:
I tried to keep this as similar to the original as possible. For instance someone might have their own subclass defined of DataFrame.